### PR TITLE
autofix: discard captures + some multi for loop support

### PR DESCRIPTION
**Pull request by travisstaloch**
this patch makes autofix add discards for unused loop and switch case captures which have curlies.  it prevents adding duplicate 'remove capture' actions by checking previous action ranges.

it removes special casing of index captures now that multi for loops have arrived.

* make getCaptureLoc() return only single Loc
* remove CaptureLocs struct which is no longer used
* add DiagnosticKind.@"switch tag capture" to handle 'inline else => |x, tag|' discards
* add test "getCaptureLoc"

### CODE REVIEW

File: src/features/code_actions.zig [hunk 1]
```diff
@@ -166,64 +164,67 @@ fn handleUnusedVariableOrConstant(builder: *Builder, actions: *std.ArrayListUnma
...
+    // search for a matching action by comparing ranges.
+    const remove_cap_loc = builder.createTextEditLoc(capture_loc, "");
+    const found_remove_capture_action = outer: for (actions.items) |item| {
+        const edit = item.edit orelse continue;
+        const changes = edit.changes orelse continue;
+        var iter = changes.iterator();
+        while (iter.next()) |entry| {
+            const t_edits = entry.value_ptr.*;
+            for (t_edits) |t_edit| {
+                if (remove_cap_loc.range.start.line == t_edit.range.start.line and
```
Comment (SuperAuguste) [inline src/features/code_actions.zig:209]: At a glance this seems HashMapable, I like the idea you mentioned where we'd pass around the map with actions :)

File: src/features/code_actions.zig [hunk 2]
```diff
@@ -166,64 +164,67 @@ fn handleUnusedVariableOrConstant(builder: *Builder, actions: *std.ArrayListUnma
...
-                "",
-            )}),
-        });
+    // look for next non-whitespace after last '|'. if its a '{' we can insert discards.
+    // this means bare loop/switch captures (w/out curlies) aren't supported.
+    var block_start = capture_loc.end + 1;
+    while (block_start < builder.handle.text.len and
+        std.ascii.isWhitespace(builder.handle.text[block_start])) : (block_start += 1)
+    {}
+    if (builder.handle.text[block_start] != '{') {
```
Comment (SuperAuguste) [inline src/features/code_actions.zig:179]: This breaks for a case like
```
    for (0..10, 0..10, 0..10) |i, j, k| i = 1;
```
but so does every other capture, definitely doesn't have to be fixed in this PR; perhaps we could open an issue for this :)

File: src/features/code_actions.zig [hunk 3]
```diff
@@ -166,64 +164,67 @@ fn handleUnusedVariableOrConstant(builder: *Builder, actions: *std.ArrayListUnma
...
-            .edit = try builder.createWorkspaceEdit(&.{builder.createTextEditLoc(
-                .{ .start = capture_locs.value.end, .end = capture_locs.loc.end - 1 },
-                "",
-            )}),
-        });
+    // look for next non-whitespace after last '|'. if its a '{' we can insert discards.
+    // this means bare loop/switch captures (w/out curlies) aren't supported.
+    var block_start = capture_loc.end + 1;
+    while (block_start < builder.handle.text.len and
+        std.ascii.isWhitespace(builder.handle.text[block_start])) : (block_start += 1)
```
Comment (SuperAuguste) [inline src/features/code_actions.zig:177]: Another edge case that we can open an issue for:
```
if (1 == 1) |a|
    //a
    {}
```
won't work as the comment isn't whitespace - I've literally never seen zig code that does this in the wild but you never know - perhaps we could even tag this issue contributor friendly and bring in some new peeps :)

File: src/features/code_actions.zig (hunk 2 above)
Comment (travisstaloch) [reply src/features/code_actions.zig:189]: created https://github.com/zigtools/zls/issues/1203

File: src/features/code_actions.zig (hunk 3 above)
Comment (travisstaloch) [reply src/features/code_actions.zig:187]: Created https://github.com/zigtools/zls/issues/1204

**Comment by travisstaloch** [comment]
addresses #689

tested with
```zig
test {
    for (0..10, 0..10, 0..10) |i, j, k| {}
    switch (union(enum) {}) {
        inline .a => |cap, tag| {},
    }
    if (null) |x| {}
    if (null) |v| {} else |e| {}
}
```
this becomes:
```zig
test {
    for (0..10, 0..10, 0..10) |i, j, k| {
        _ = k;
        _ = j;
        _ = i;
    }
    switch (union(enum) {}{}) {
        inline .a => |cap, tag| {
            _ = tag;
            _ = cap;
        },
    }
    if (null) |x| {
        _ = x;
    }
    if (null) |v| {
        _ = v;
    } else |e| {
        _ = e;
    }
}
```

**Comment by leecannon** [comment]
This looks good :+1:

**Review by leecannon — APPROVED** [review]

**Comment by travisstaloch** [comment]
interesting note: just tested a release-fast build and the discard order is reversed back to source order. 

```zig
test {
    for (0..10, 0..10, 0..10) |i, j, k| {
        _ = i;
        _ = j;
        _ = k;
    }
...
```

glad to see they're in source order. not sure why they were reversed in the debug build.

**Pull request merged**